domain/state.go

```diff
@@ -99,6 +100,84 @@ func (st *StateBase) Prepare(query string, typeSamples ...any) (*sqlair.Statemen
 	return stmt, nil
 }
 
+// RunTx executes the closure function within the scope of a transaction.
+// The closure is passed a TxContext that can be passed on to state
+// functions, so that they can perform work within that same transaction.
+// The closure will be retried according to the transaction retry semantics,
+// if the transaction fails due to transient errors.
+// The closure should only be used to perform state changes and must not be used
+// to execute queries outside of the state scope.
+func (st *StateBase) RunTx(ctx context.Context, fn func(TxContext) error) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Annotate(err, "getting database")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// The txContext is created with the transaction and passed to the
+		// closure function. This ensures that the transaction is always
+		// available to the closure. Once the transaction is complete, the
+		// transaction is removed from the txContext. This is to prevent
+		// the transaction from being used outside of the transaction scope.
+		// This will prevent any references to the sqlair.TX from being held
+		// outside of the transaction scope.
+
+		// TODO (stickupkid): The txContext can be pooled on the StateBase to
+		// reduce the number of allocations. Attempting to push the tx into
+		// the context would prevent that as a viable option.
+		txCtx := &txContext{
+			Context: ctx,
+			t:       tx,
+		}
+		defer txCtx.removeTx()
+
+		return fn(txCtx)
+	})
+}
+
+// TxStateBase is an interface that provides a method for executing a closure
+// within the scope of a transaction.
+type TxStateBase interface {
+	// RunTx executes the closure function within the scope of a transaction.
+	// The closure is passed a TxContext that can be passed on to state
+	// functions, so that they can perform work within that same transaction.
+	// The closure will be retried according to the transaction retry semantics,
+	// if the transaction fails due to transient errors.
+	// The closure should only be used to perform state changes and must not be
+	// used to execute queries outside of the state scope.
+	RunTx(ctx context.Context, fn func(TxContext) error) error
+}
+
+// Run executes the closure function using the provided TxContext as
+// the transaction context.
+// It is expected that the closure will perform state changes within the
+// transaction scope.
+// Any errors returned from the closure are coerced into a standard error
+// to prevent sqlair errors from being returned to the Service layer.
+func Run(ctx TxContext, fn func(context.Context, *sqlair.TX) error) error {
```

Comment:
Note that only convention forces us to use `Run` to wrap calls in the state layer. We will have to decide whether this is a compulsory convention.

Reply:
Correct. We could have a function, that can drop the closure. The downside is you'd have to coerce the error yourself before returning back to the service.

```go
func GetTx(ctx TxContext) (*sqlair.Tx, error)
```

Reply:
Let's keep Run. It (with the mutex) gives us some guarantees that I like.